### DISCUSSION TITLE
Variable Management in Github

### DISCUSSION THREAD

**Discussion by jnus** [community/community#25755] [Actions]
<p>Currently moving from an on-premise pipeline composed of Bitbucket, TeamCity and Octopus Deploy to Github and loving having a single platform where I can define everything as code - so far so good. But to my surprise and coming from Octopus Deploy, variable management is really a big pain point and pretty much a deal breaker for many teams. And I’m a little puzzled as why this is not discussed more in the community (I may have missed it), making me thing we might have missed something.</p>
<p>The main features I’m missing from Octopus Deploy</p>
<ul>
<li><a href="https://octopus.com/docs/projects/variables/library-variable-sets" rel="noopener nofollow ugc">Global variable sets scoped to environments</a></li>
<li><a href="https://octopus.com/docs/projects/variables" rel="noopener nofollow ugc">Repo variables scoped to environments</a></li>
<li>
<a href="https://octopus.com/docs/projects/variables/variable-substitutions" rel="noopener nofollow ugc">Variable substitution concept</a>, so you have convention based variables (e.g. environment name substituted into URL).</li>
</ul>
<p>What we are currently doing in our CICD pipeline in GH is explicitely declaring environment variables for each and every environment (job), using repo and org. secrets for the bits we do not want to expose. This gets even worse when you have adhoc workflows for registrating e.g. idenity metadata for API’s and clients, therefor duplicating the exact same variables into additional workflows.</p>
<p>Questions:</p>
<ol>
<li>Is this really the only approach at the moment?</li>
<li>Are there any active development (or plans) in terms of an actual variable management framework?</li>
<li>Are there any active discussions regarding this, we could chip into?</li>
<li>Any community contributions/actions trying to solve this issue worth looking into?</li>
</ol>
<p>Any information is greatly appreciated!</p>

**Comment by spol-vt** [comment]
<p>Those features, at least when it comes to managing <code>secrets</code>, are available, but only with GitHub Enterprise Cloud.</p>

**Comment by lirantal** [comment]
<p><a class="mention" href="/spol-vt">@spol-vt</a> could you provide a link with reading material in the docs to that extended variable management capabilities?</p>
<p><a class="mention" href="/jnus">@jnus</a> now that composed actions have much more flexibility (announced a few weeks back) - I am wondering if you would maybe want to create a sort of repository of configuration where all the data lives in JS files, and use an action that works based on that. Kind of like in the concept of GitOps. You’ll manage those variables simply by editing those JSON files.</p>

**Comment by spol-vt** [comment]
<p><a class="mention" href="/lirantal">@lirantal</a> <a href="https://docs.github.com/en/actions/reference/environments" class="inline-onebox" rel="noopener nofollow ugc">Environments - GitHub Docs</a></p>
<blockquote>
<p>Environment protection rules and environment secrets are only available on private repositories with GitHub Enterprise Cloud. For more information about how you can try GitHub Enterprise Cloud for free, see “[Setting up a trial of GitHub Enterprise Cloud]”</p>
</blockquote>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://docs.github.com/en/actions/learn-github-actions/sharing-workflows-with-your-organization#sharing-secrets-within-an-organization">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">

      <a href="https://docs.github.com/en/actions/learn-github-actions/sharing-workflows-with-your-organization#sharing-secrets-within-an-organization" target="_blank" rel="noopener nofollow ugc">GitHub Docs</a>
  </header>

  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/79195663/181093065-ccb3a3e2-cfa2-40ad-ba2d-830f3c22c90e.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://docs.github.com/en/actions/learn-github-actions/sharing-workflows-with-your-organization#sharing-secrets-within-an-organization" target="_blank" rel="noopener nofollow ugc">Sharing workflows with your organization - GitHub Docs</a></h3>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<blockquote>
<p>You can centrally manage your secrets within an organization, and then make them available to selected repositories. This also means that you can update a secret in one location, and have the change apply to all repository workflows that use the secret.</p>
</blockquote>
<p>There is nothing for classic clear txt vars, just secrets, however since it is basically same difference, you can use it for that purpose. With Enterprise Could one can create Org level, Repo level and then Environment level secrets + there is permission management around access to secrets, etc. Functionality is not as extensive as in products mentioned by <a class="mention" href="/jnus">@jnus</a>, but there is something that goes towards it.</p>

**Comment by ThomasReyskens** [comment]
<p>The secrets per environment are nice if you an enterprise plan. The thing I miss the most is that you can’t have variables that are readable… Why only have secrets? :sweat_smile:</p>

**Comment by jnus** [comment]
<p><a class="mention" href="/spol-vt">@spol-vt</a>  - So some of the feature set is partly there, in terms of secrets, but not on an organization level though where you can’t scope to environments. We share a ton of variables today via Octopus Deploy, scoped for the various environments, ready to use. This could be resolved in others ways than inheritage (submoduling, npm/nupkg etc), so I’m not seeing this as the biggest challange. Not having a unified and context specific ‘view’ of you variables in you workflow(s), in the really big pain point, where you have to declare all environment specific varibles per environment prior to deploying.</p>
<p><a class="mention" href="/lirantal">@lirantal</a> Having all variables as secrets will not be easy to maintain, plus I really applaud having everything in vcs.</p>
<p><a class="mention" href="/thomasreyskens">@ThomasReyskens</a>  - It’s a good question - why only have the options to scope secrets per environment and not clear text? My guess is that the GH team also loves having everything in vsc, including configuration and giving the option to save clear text variables on the GH controlplane, violates this principal. A variable framework should definitely based on having the configuration in vsc and having secrets outside, is just a necessary evil to keep sensitive information safe.</p>

**Comment by jnus** [comment]
<p><a class="mention" href="/lirantal">@lirantal</a> - regarding composed actions, that might be an idea going forward. Depending on what kind of initiative is going on around the community, I might start to build a variable framework based on the ideas of Octopus Deploy - e.g. variables in a json structure, scopes to environment/workflow/jobs/steps, supports variable substitution and context aware depending on the environment you are deploying to, workflow etc. But I really want to make sure this is not on the backlog for GH, as I want to use vanilla GH if possible :expressionless:</p>

**Comment by lirantal** [comment]
<p>Makes sense, yeah :slight_smile:</p>

**Comment by dorner** [comment]
<p>I ended up building a “pre-processing” tool for GHA workflows tailored to my company. This included variables, an equivalent to YAML anchors and pre-populating jobs with common steps. Downside is we have to remember to re-run this process every time we change the input workflow, but it makes it much simpler to manage.</p>

**Comment by jnus** [comment]
<p><a class="mention" href="/dorner">@dorner</a> interesting way to go about it. If you are allowed to share any details about your current implementation (publicly/privately), that would be greatly appreciated for inspirational purposes.</p>
<p>Still really puzzled to why this is not on the backlog for the GH team, with all the focus Microsoft is giving GitHub, as this is a big big pain point with larger deployment setups.</p>

**Comment by dorner** [comment]
<p><a class="mention" href="/jnus">@jnus</a> you can see a cleaned-up version of it here: <a href="https://github.com/dorner/gha-config" rel="noopener nofollow ugc">https://github.com/dorner/gha-config</a> This version probably doesn’t work per se because I scrubbed some of it, but that’s the general approach.</p>

**Comment by jnus** [comment, marked as answer]
<p>So I’ve been working on an action, json-variables, for this particular GH issue, and while its still work in progress before I publish it, I thought I might share it here and maybe get a little input.</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/jnus/json-variables">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/jnus/json-variables" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/2075403/181093073-82185f38-7391-4fd9-bd5d-e66a6083c886.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/d/6/d6f56e7c39c84b428b648045998231a10baa09a6_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/d/6/d6f56e7c39c84b428b648045998231a10baa09a6_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/d/6/d6f56e7c39c84b428b648045998231a10baa09a6.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/d/6/d6f56e7c39c84b428b648045998231a10baa09a6_2_10x10.png"></div>

<h3><a href="https://github.com/jnus/json-variables" target="_blank" rel="noopener nofollow ugc">GitHub - jnus/json-variables</a></h3>

  <p>Contribute to jnus/json-variables development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Let me know what you think.</p>

**Comment by jnus** [comment]
<p>Fyi - v1.0 of the action just released → <a href="https://github.com/marketplace/actions/json-variables" class="inline-onebox" rel="noopener nofollow ugc">json-variables · Actions · GitHub Marketplace · GitHub</a></p>
